docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,319 @@
+
+A fundamental change compared to the current sync mechanism is that one source path may by sync'ed to multiple destinations in the future.
+
+### Assumptions
+
+Builders know where to put static files in the container.
+This information needs to be pulled from the builders and made available in Skaffold.
+Currently, we know how to do this for Docker artifacts (@r2d4 suggested this in #1166).
+Whether jib and bazel builders can provide that information too, is unclear at the moment.
+
+### Interface changes
+
+There are at least two relevant changes
+
+1. When builders are queried for their dependencies, they must provide a list of destinations together with local source paths. The signature of `Builder.GetDependencies` (pkg/skaffold/build/build.go) therefore needs to change to:
+   ```go
+   type Builder interface {
+       Labels() map[string]string
+
+       Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+
+       // OLD
+       // DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) ([]string, error)
+       // NEW
+       DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) (map[string][]string, error)
+   }
+   ```
+   
+   So far, `DependenciesForArtifact` returned a list of local paths (relative to the working directory of the artifact) that are input files for the container.
+   Now, `DependenciesForArtifact` serves two purposes. As before, the map keys are relative local paths of input files for the container.
+   In addition, the map values list all absolute remote destination paths where the input files are placed in the container.
+   This has to be a list, because a single input file may be copied to several locations in the container.
+   
+   Thus, `DependenciesForArtifact` serves two purposes: firstly, it lists all input files to watch for changes.
+   Secondly, it provides default sync destinations for changed or deleted files.
+   Note that this does not mean, that every file will be sync'ed. If a file is entitled for syncing depends on the user-provided sync rules.
+   
+   For example, given the above Dockerfile, it will return something like
+   ```go
+   map[string][]string{
+    "static/page1.html": []string{"/var/www/page1.html"},
+    "static/assets/style.css": []string{"/var/www/assets/style.css"},
+    "index.html": []string{"/var/www/index.html"},
+    "nginx.conf": []string{"/var/nginx.conf", "/etc/nginx.conf"},
+   }
+   ```
+   
+2. To support additional sync maps from builders, the builder interface will need an additional method `SyncMap`:
+   ```go
+   type Builder interface {
+       Labels() map[string]string
+
+       Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+
+       DependenciesForArtifact(ctx context.Context, artifact *latest.Artifact) (map[string][]string, error)
+    
+       // ADDED
+       SyncMap() map[string]string
```

Comment:
technically this should be map[string][]string as well (`SyncRule`) to match copy semantics

Reply:
@balopat @briandealwis So it just occurred to me that in my original proposal, `GetDependencies` and `SyncMap` were answering different questions. `GetDependencies` would answer _how_ to sync files, whereas `SyncMap` would answer _what_ files to sync.

For the reasons outlined above, we want to keep `GetDependencies` unchanged and `SyncMap` will get the semantics of _how_ to sync. However, this leaves open the question of _what_ to sync. As @balopat outlined above, this magic jib sync will just be mentioned and not be further detailed by this design proposal. However, I wanted to highlight that this is now out of scope for this design proposal. Correct?

Reply:
+1 for out of scope now

Reply:
yes!